File: R/ZigZag.R
#
#   TTR: Technical Trading Rules
#
#   Copyright (C) 2007-2013  Joshua M. Ulrich
#
#   This program is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 2 of the License, or
#   (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

#'Zig Zag
#'
#'Zig Zag higlights trends by removing price changes smaller than \code{change}
#'and interpolating lines between the extreme points.
#'
#'The Zig Zag is non-predictive.  The purpose of the Zig Zag is filter noise
#'and make chart patterns clearer.  It's more a visual tool than an indicator.
#'
#'@aliases ZigZag zigzag
#'@param HL Object that is coercible to xts or matrix and contains either a
#'High-Low price series, or a Close price series.
#'@param change Minimum price movement, either in dollars or percent (see
#'\code{percent}).
#'@param percent Use percentage or dollar change?
#'@param retrace Is \code{change} a retracement of the previous move, or an
#'absolute change from peak to trough?
#'@param lastExtreme If the extreme price is the same over multiple periods,
#'should the extreme price be the first or last observation?
#'@return A object of the same class as \code{HL} or a vector (if
#'\code{try.xts} fails) containing the Zig Zag indicator.
#'@note If High-Low prices are given, the function calculates the max/min using
#'the high/low prices.  Otherwise the function calculates the max/min of the
#'single series.
#'@author Joshua Ulrich
#'@references The following site(s) were used to code/document this
#'indicator:\cr
#'\url{http://www.fmlabs.com/reference/default.htm?url=ZigZag.htm}\cr
#'\url{http://www.linnsoft.com/tour/techind/zigzag.htm}\cr
#'\url{http://www.linnsoft.com/tour/techind/zigosc.htm}\cr
#'\url{http://www.equis.com/Customer/Resources/TAAZ/?c=3&p=127}\cr
#'\url{http://stockcharts.com/school/doku.php?id=chart_school:technical_indicators:zigzag}\cr
#'@keywords ts
#'@examples
#'
#' ## Get Data and Indicator ##
#' data(ttrc)
#' zz <- ZigZag( ttrc[,c("High", "Low")], change=20 )
#'
#'@export
"ZigZag" <- 
function( HL, change=10, percent=TRUE, retrace=FALSE, lastExtreme=TRUE ) {

  # Zig Zag Indicator
  # Adapted from Alberto Santini's code

  HL <- try.xts(HL, error=as.matrix)
  HL.na <- naCheck(HL,0)
  
  # Calculation if HL series is given
  if(NCOL(HL)==2) {
    high  <- HL[HL.na$nonNA,1]
    low   <- HL[HL.na$nonNA,2]
  } else

  # Calculation if price vector is given
  if(NCOL(HL.na)==1) {
    high  <- HL[HL.na$nonNA]
    low   <- HL[HL.na$nonNA]
  } else

  stop("Price series must be either High-Low, or Univariate")
  
  # Initialize necessary parameters
  nn <- NROW(HL.na$nonNA)
  zz <- rep(0, nn)
  if(percent) {
    change <- change/100
  }

  # Call Fortran routine
  zz <- .Fortran("zigzag", iha = as.double( high ),
                           ila = as.double( low ),
                           la  = as.integer( nn ),
                           ch  = as.double( change ),
                           pct = as.integer( percent ),
                           rtr = as.integer( retrace ),
                           lex = as.integer( lastExtreme ),
                           zz  = as.double( zz ),
                           PACKAGE = "TTR",
                           DUP = TRUE )$zz
  
  # Interpolate results
  zz <- ifelse( zz==0, NA, zz )
  zz <- approx( zz, xout=1:nn )$y

  # Prepend NAs from original data
  zz <- c( rep( NA, HL.na$NAs ), zz ) 

  reclass( zz, HL )
}

File: pkg/src/zigzag.f
c
c   TTR: Technical Trading Rules
c
c   Copyright (C) 2007-2013  Joshua M. Ulrich
c
c   This program is free software: you can redistribute it and/or modify
c   it under the terms of the GNU General Public License as published by
c   the Free Software Foundation, either version 2 of the License, or
c   (at your option) any later version.
c
c   This program is distributed in the hope that it will be useful,
c   but WITHOUT ANY WARRANTY; without even the implied warranty of
c   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
c   GNU General Public License for more details.
c
c   You should have received a copy of the GNU General Public License
c   along with this program.  If not, see <http://www.gnu.org/licenses/>.
c

c
c     Zig Zig
c
c     iha   : input array of high prices
c     ila   : input array of low prices
c     la    : length of arrays
c     ch    : minimum change to cause a zig/zag
c     pct   : logical, % (1) or $ (0) change
c     rtr   : logical, retrace (1) or absolute (0) change
c     lex   : logical, last (1) or first (0) extreme value
c     zz    : Zig/Zag vector
c
c     refpos: Reference (first) price position
c     refval: Reference (first) price value
c     infpos: Inflection (second) price position
c     infval: Inflection (second) price value
c     *min  : [L]ocal and [E]xtreme minimums
c     *max  : [L]ocal and [E]xtreme maximums
c
      subroutine zigzag(iha, ila, la, ch, pct, rtr, lex, zz)
      implicit none

      integer la, pct, rtr, lex, i, refpos, infpos, sig
      double precision iha(la), ila(la), zz(la)
      double precision refval, infval, ch
      double precision lmin, lmax, emin, emax
c
c     Initialize values
c
      refval = (iha(1) + ila(1)) / 2
      infval = (iha(2) + ila(2)) / 2
      refpos = 1
      infpos = 2
      sig = 0
c
c     Begin Loop
c
      do 10 i=2,la

      if( pct .EQ. 1 ) then
c     If % change given (absolute move)
          emin = infval * ( 1 - ch )
          emax = infval * ( 1 + ch )
      else
c     If $ change given (only absolute moves make sense)
          emin = infval - ch
          emax = infval + ch
      endif
c
c     Find local maximum and minimum
c
      lmax = MAX( infval, iha(i) )
      lmin = MIN( infval, ila(i) )
c
c     Find first trend
c
      if( sig .EQ. 0 ) then
          if( rtr .EQ. 1 ) then
c         Retrace prior move
              if( infval .GE. refval ) then
                  sig = 1
              else
                  sig = -1
              endif
          else
c         Absolute move
              if( lmin .LE. emin ) then
c             Confirmed Downtrend
                  sig = -1
              endif
              if( lmax .GE. emax ) then
c             Confirmed Uptrend
                  sig = 1
              endif
          endif
      endif
c
c     Downtrend
c
      if( sig .EQ. -1 ) then
c
c         New Minimum
c
          if( ila(i) .EQ. lmin ) then
c             Last Extreme
              if( lex .EQ. 1 ) then
                  infval = ila(i)
                  infpos = i
              else
c             First Extreme
                  if( ila(i) .NE. ila(i-1) ) then
                      infval = ila(i)
                      infpos = i
                  endif
              endif
          endif
c
c         Retrace prior move
c
          if( rtr .EQ. 1 ) then
              emax = infval + ((refval - infval) * ch)
          endif
c
c         Trend Reversal
c
          if( iha(i) .GE. emax ) then
              zz(refpos) = refval
              refval = infval
              refpos = infpos
              infval = iha(i)
              infpos = i
              sig = 1
              GO TO 10
          endif
      endif
c
c     Uptrend
c
      if( sig .EQ. 1 ) then
c
c         New Maximum
c
          if( iha(i) .EQ. lmax ) then
c             Last Extreme
              if( lex .EQ. 1 ) then
                  infval = iha(i)
                  infpos = i
              else
c             First Extreme
                  if( iha(i) .NE. iha(i-1) ) then
                      infval = iha(i)
                      infpos = i
                  endif
              endif
          endif
c
c         Retrace prior move
c
          if( rtr .EQ. 1 ) then
              emin = infval - ((infval - refval) * ch)
          endif
c
c         Trend Reversal
c
          if( ila(i) .LE. emin ) then
              zz(refpos) = refval
              refval = infval
              refpos = infpos
              infval = ila(i)
              infpos = i
              sig = -1
              GO TO 10
          endif
      endif

   10 continue
c
c     Set final values
c     
      zz(refpos) = refval
      zz(infpos) = infval
      end
